baseplate_test.go

```diff
@@ -145,6 +150,64 @@ func TestServe(t *testing.T) {
 	}
 }
 
+type timestampCloser struct {
+	ts []time.Time
+}
+
+func (c *timestampCloser) Close() error {
+	c.ts = append(c.ts, time.Now())
+	return nil
+}
+
+func TestServeClosers(t *testing.T) {
+	t.Parallel()
+
+	store := newSecretsStore(t)
+	defer store.Close()
+
+	bp := baseplate.NewTestBaseplate(baseplate.Config{StopTimeout: testTimeout}, store)
+
+	pre := &timestampCloser{}
+	post := &timestampCloser{}
+
+	args := baseplate.ServeArgs{
+		Server:       newWaitServer(t, bp, time.Millisecond),
+		PreShutdown:  []io.Closer{pre},
+		PostShutdown: []io.Closer{post},
+	}
+
+	ch := make(chan error)
+
+	go func() {
+		// Run Serve in a goroutine since it is blocking
+		ch <- baseplate.Serve(context.Background(), args)
+	}()
+
+	time.Sleep(time.Millisecond)
+	p, err := os.FindProcess(syscall.Getpid())
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	p.Signal(os.Interrupt)
+	<-ch
+
+	if len(pre.ts) != 1 {
+		t.Fatalf("Unexpected number of PreShutdown calls: expected 1, got %v", len(pre.ts))
+	}
+	if len(post.ts) != 1 {
+		t.Fatalf("Unexpected number of PostShutdown calls: expected 1, got %v", len(post.ts))
+	}
+
+	if !pre.ts[0].Before(post.ts[0]) {
+		t.Fatalf(
```

Comment:
I'd prefer not to overly use Fatal over Error in tests. The previous ones are fine: We can't continue the test if we can't trigger graceful shutdown, getting `pref.ts[0]` could panic if `len(pre.ts) < 1`, but here this is just an error, not some thing fatal.